### PR TITLE
bpo-29708: allow to force hash-based pycs

### PR DESCRIPTION
**Pull request by bmwiedemann**
allow to force hash-based pycs
using the well-established `SOURCE_DATE_EPOCH` environment variable
to allow for reproducible builds of python packages.

See https://reproducible-builds.org/ for why this is good.

This implements variant 1 from @brettcannon

note: not tested yet

https://bugs.python.org/issue29708

### CODE REVIEW

File: Doc/library/py_compile.rst
Comment (brettcannon) [inline Doc/library/py_compile.rst]: This should also be mentioned in the main documentation. In other words, if the "version changed" note went away the docs should still cover what they need to cover. So this should say something like "Detecting the :envvar:`SOURCE_DATE_EPOCH` environment variable to force `CHECKED_HASH` was added."

File: Lib/test/test_py_compile.py
Comment (brettcannon) [inline Lib/test/test_py_compile.py]: I would look at https://github.com/python/cpython/blob/9f1b7b93f5f0ef589e7b272e127cacf4ce5d23f1/Lib/test/test_py_compile.py#L125 as an example on how to make this test a little bit more robust in the face of `.pyc` files changing.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Just a request to expand the documentation and tweak the test to follow what is already there for hash-based .pyc files.

**Comment by bmwiedemann** [comment]
I have made the requested changes; please review again.

**Review by brettcannon — APPROVED** [review]
Just waiting until Friday to merge to give people enough time to comment on the proposed solution. Otherwise the PR LGTM!

**Comment by brettcannon** [comment]
Thanks for your patience with this issue, @bmwiedemann !

**Pull request merged**
